pokemon.go

```diff
@@ -53,6 +67,137 @@ func GetPokemon(natdex uint16) Pokemon {
 	panic(fmt.Sprintf("unknown Pokedex number %v\n", natdex))
 }
 
+// Creates a new Pokemon given its national dex number
+func GeneratePokemon(args ...interface{}) Pokemon {
+	// Mandatory parameters
+	var natdex uint16
+
+	// Optional parameters
+	var level uint8 = 1
+	var totalExp uint // used to compute level if provided
+	var ivs [6]uint8 = [6]uint8{0, 0, 0, 0, 0, 0}
+	var evs [6]uint8 = [6]uint8{0, 0, 0, 0, 0, 0}
+	var nature *Nature = GetNatureTable()["hardy"] // this nature provides no bonuses/debuffs
+
+	if len(args) < 2 {
+		panic("Not enough parameters.")
+	}
+
+	for i, p := range args {
+		switch i {
+		case 0: // natdex
+			param, ok := p.(uint16)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type uint16, got %v", i, reflect.TypeOf(p)))
+			}
+			natdex = param
+		case 1: // level or total exp gained
+			switch p.(type) {
+			case uint8: // level
+				level = p.(uint8)
+				totalExp = computeExpFromLevel(level, getGrowthRateFromDexNumber(natdex))
+			case uint: // total experience
+				totalExp = p.(uint)
+				level = computeLevelFromExp(totalExp, getGrowthRateFromDexNumber(natdex))
+			default:
+				panic(fmt.Sprintf("Parameter %d: expected type uint8 or uint, got %v", i, reflect.TypeOf(p)))
+			}
+		case 2: // ivs
+			param, ok := p.([6]uint8)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type [6]uint8, got %v", i, reflect.TypeOf(p)))
+			}
+			ivs = param
+		case 3: //evs
+			param, ok := p.([6]uint8)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type [6]uint8, got %v", i, reflect.TypeOf(p)))
+			}
+			evs = param
+		case 4: // nature
+			param, ok := p.(*Nature)
+			if !ok {
+				panic(fmt.Sprintf("Parameter %d: expected type Nature, got %v", i, reflect.TypeOf(p)))
+			}
+			nature = param
+		}
+	}
```

Comment:
Variadic interface for this case is really ambiguous for the user. Couple of solutions to make this more clear:
- `map[string]interface{}`: gives a name to each key rather than the index]
- custom struct type: good because has explicit field names that the user can see, but difficult to test for zero values (e.g. passing in 0 vs. not passing in any value)
- [further optional parameter techniques](https://stackoverflow.com/questions/2032149/optional-parameters-in-go): probably a bit too complex for our needs, but shows composability in Go

[This article](https://petomalina.medium.com/dealing-with-optional-parameters-in-go-9780f9bfbd1d) describes better approaches to optional parameters.